platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/adapter/ProtocolToModelAdapter.java

```diff
@@ -187,17 +187,29 @@ private static <T> Class<?>[] getModelContractInterfaces(Class<T> targetType, Ob
     }
 
     private static <T> Map<String, Class<?>> getPotentialModelContractSubInterfaces(Class<T> targetType) {
-        Annotation[] annotations = targetType.getAnnotations();
-        Map<String, Class<?>> result = new HashMap<>();
-        for (Annotation annotation : annotations) {
-            if (annotation instanceof ToolingModelContract) {
-                Class<?>[] classes = ((ToolingModelContract) annotation).subTypes();
-                for (Class<?> clazz : classes) {
-                    result.put(clazz.getName(), clazz);
+        HashMap<String, Class<?>> result = new HashMap<>();
+        getPotentialModelContractSubInterfaces(targetType, new HashSet<Class<?>>(), result);
+        return result;
+    }
+
+    private static <T> void getPotentialModelContractSubInterfaces(
+        Class<T> targetType,
+        Set<Class<?>> visited,
+        Map<String, Class<?>> result
+    ) {
+        boolean isNew = visited.add(targetType);
+        if (isNew) {
+            Annotation[] annotations = targetType.getAnnotations();
+            for (Annotation annotation : annotations) {
+                if (annotation instanceof ToolingModelContract) {
+                    Class<?>[] classes = ((ToolingModelContract) annotation).subTypes();
+                    for (Class<?> clazz : classes) {
+                        result.put(clazz.getName(), clazz);
+                        getPotentialModelContractSubInterfaces(clazz, visited, result);
```

Comment:
❓ Could this lead to stackoverflows if there's a cycle in involved the interfaces? It would be better to not revisit already visited types.

Reply:
It's actually already handled...